🐛 iamauth: reconcile mappings as desired state (PCP-2390)

wcrum opened the pull request:
## Summary

Fixes `PCP-2390` — CAPA does not remove roles/users from `kube-system/aws-auth` when they are removed from `AWSManagedControlPlane.spec.iamAuthenticatorConfig`. Customer-facing security bug reported by an enterprise customer running Palette-provisioned EKS clusters (parent trackers `SUS-243`, `SCS-1324`). Pending since 2024-01.

## Why the previous PR was insufficient

This PR replaces the earlier implementation on this branch entirely (force-push after retargeting from `spectro-4.6.b` → `spectro-master`). The earlier attempt had two critical defects:

1. **Cluster-breaking regression on ConfigMap backend** — the earlier `MapRoles` cleared `authConfig.RoleMappings = []` before appending, but the reconciler calls `MapRole(nodeRoleMapping)` for each worker role **before** `MapRoles(iamCfg.RoleMappings)`. Result: node roles get wiped out on every reconcile → EKS workers lose kubelet auth.
2. **No-op on CRD backend** — the earlier `MapRoles` / `MapUsers` on the CRD backend just looped over the existing single-item `MapRole` / `MapUser`, which only *append*. Stale `IAMIdentityMapping` CRs were never deleted; the reported bug still reproduced for CRD-backend users.

## Design

Full internal Solution Design: <https://spectrocloud.atlassian.net/wiki/spaces/ENGINEERIN/pages/3973152777> (SSO-gated).

Approach:

- New `ReconcileMappings(roles, users)` method on `AuthenticatorBackend`. Set-based semantics: input is the full desired state.
- Reconciler composes the desired role set once (**node roles ∪ user-configured `iamCfg.RoleMappings`**) and calls `ReconcileMappings` exactly once. The old code called `MapRole` per node role, then `MapRole` per user role, then `MapUser` per user — a pattern that is architecturally incapable of expressing "delete this entry".
- ConfigMap backend replaces `mapRoles` / `mapUsers` wholesale (`saveAuthConfig` deletes the yaml keys when the slice is empty).
- CRD backend list-diffs CAPA-managed `IAMIdentityMapping` CRs (identified by `GenerateName: "capa-iamauth-"` prefix), creates missing entries, and deletes stale ones. Out-of-band CRs (no CAPA prefix) are never touched.
- `MapRole` / `MapUser` remain on the interface (deprecated) for API back-compat with any external consumers. The internal reconciler no longer calls them.

## Upstream status

Not fixed upstream. Upstream bug [kubernetes-sigs/cluster-api-provider-aws#4740](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/4740) auto-closed 2025-10-28 as "Not Planned" by triage bot. Two prior upstream PRs by Will Crum ([#4741](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4741) in 2024-04, [#4919](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4919) in 2025-08) both closed without merge. This lands as a permanent Spectro fork carry.

## Definition of Done

- [x] **`make lint` (targeted, iamauth) passes with zero warnings vs `spectro-master`.**
  Baseline (spectro-master, iamauth only) has zero warnings; branch matches. Full `make lint` fails against unrelated webhook_test.go typecheck issues that pre-exist on `spectro-master` and are not touched by this PR.
- [x] **`go test -race -cover ./pkg/cloud/services/iamauth/...` passes on this branch with `-race` clean.**
  9 test functions pass (`TestAddRoleMappingCM`, `TestAddUserMappingCM`, `TestReconcileMappingsCM` [6 sub-cases], `TestAddRoleMappingCRD`, `TestAddUserMappingCRD`, `TestReconcileMappingsCRD` [4 sub-cases], `TestReconcileMappingsCRDIdempotent`, `TestReconcileMappingsCRDInvalidARN`, `TestReconcileIAMAuth`). No race warnings.
- [x] **All backend-level scenarios from the design spec have a corresponding test.**
  `TestReconcileMappingsCM` covers empty-desired-set-clears, add, replace-removes-stale, node-role preservation, idempotency, invalid-ARN no-mutation. `TestReconcileMappingsCRD` covers delete stale CAPA CR, preserve out-of-band CR, node-role preserved, idempotency. `TestReconcileMappingsCRDIdempotent` asserts exact CR count after two calls. `TestReconcileMappingsCRDInvalidARN` asserts pre-existing CR untouched on validation failure.
- [x] **Coverage on `pkg/cloud/services/iamauth/` does not decrease.**
  Baseline: 65.9% of statements. This branch: **70.3%** of statements. +4.4 pp.
- [x] **Mock file regenerated.**
  `go generate ./pkg/cloud/services/iamauth/...` executed; the mock is for the AWS `IAMAPI` interface (aws-sdk-go), not for our internal `AuthenticatorBackend`, so no mock diff was produced. `AuthenticatorBackend` is used directly (not through mocks) — the ConfigMap and CRD backend tests exercise real backend implementations against a fake `client.Client`.
- [x] **Branch rebased on `origin/spectro-master`.**
  Two clean commits, no `spectro-4.6.b` artifacts:
  ```
  6c729aa50 test(iamauth): cover ReconcileMappings desired-state semantics (PCP-2390)
  d41028b2a fix(iamauth): reconcile mappings as desired state (PCP-2390)
  ```
- [x] **PR base retargeted to `spectro-master`.**
- [x] **PR body updated** — this text.
- [x] **Bug reproduction rationale documented.** The PR replaces the previous append-only path with a single `ReconcileMappings` call driven by the reconciler-composed desired set. A user removed from `iamAuthenticatorConfig` is now absent from the desired set → the ConfigMap backend rewrites `mapUsers` without them; the CRD backend deletes the stale CAPA-managed CR. Node roles are guaranteed present because the reconciler unions them into the desired set before the single backend call — architecturally impossible for a subsequent call to wipe them. E2E on a real EKS cluster deferred to internal QA.
- [ ] **Approver assigned per OWNERS.** (Requires reviewer action on GitHub.)

## Testing

- **Unit tests:** 11 new test cases across `configmap_test.go` and `crd_test.go`:
  - `TestReconcileMappingsCM` — 6 sub-cases (empty-set clears, add, replace-removes-stale, node-role preservation, idempotency, invalid-ARN no-mutation)
  - `TestReconcileMappingsCRD` — 4 sub-cases (delete stale CAPA CR, preserve out-of-band CR, node-role preservation, idempotency)
  - `TestReconcileMappingsCRDIdempotent` — asserts exact list size after two identical calls
  - `TestReconcileMappingsCRDInvalidARN` — asserts pre-existing CR untouched on validation failure
- **Coverage:** 65.9% → **70.3%** of statements on `pkg/cloud/services/iamauth/`.
- **Reconciler-level test:** deferred to internal QA E2E. The composition logic (`desiredRoles := nodeRoles ∪ iamCfg.RoleMappings`) is a 5-line union in `reconcile.go` that feeds directly into the covered backend method; the bug in the previous PR came from separate `MapRole` calls competing with a subsequent `MapRoles` that pre-cleared state — a shape the new architecture cannot express.
- **`-race`:** clean across all tests in the package.

## Fixes

- Fixes `PCP-2390`
- Unblocks `SUS-243`, `SCS-1324`

wcrum committed:
fix: remove previous roles and users

spectro-prow commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/spectrocloud/cluster-api-provider-aws/pull/890#" title="Author self-approved">wcrum</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign  after the PR has been reviewed.
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=spectrocloud%2Fcluster-api-provider-aws).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/spectrocloud/cluster-api-provider-aws/blob/spectro-master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

wcrum committed:
fix: add unit tests

ChellaVigneshKP committed:
fix(iamauth): reconcile mappings as desired state (PCP-2390)

Change the AuthenticatorBackend contract so aws-iam-authenticator
configuration is applied as a desired state instead of append-only.

The previous implementation on both backends only added entries — a
user or role removed from iamAuthenticatorConfig was never cleaned up
(SUS-243 / SCS-1324). Compose the desired role set at the reconciler
(node roles union iamCfg.RoleMappings) and pass the full set to a new
ReconcileMappings method on each backend in a single call:

  * configMapBackend: replaces mapRoles and mapUsers wholesale.
  * crdBackend: list-diffs CAPA-managed IAMIdentityMapping CRs by
    the "capa-iamauth-" GenerateName prefix, creates missing, and
    deletes stale. Out-of-band CRs (no CAPA prefix) are untouched.

Deprecate MapRole / MapUser on the interface but keep them for API
back-compat. The reconciler no longer calls them.

ChellaVigneshKP committed:
test(iamauth): cover ReconcileMappings desired-state semantics (PCP-2390)

Add table-driven backend tests for the new ReconcileMappings method.

configMapBackend (TestReconcileMappingsCM, 6 sub-cases):
  * empty desired set on populated backend removes everything
  * add user to empty backend
  * replace {alice,bob} with {alice} removes bob
  * node role preserved when included in desired set
  * idempotency across repeated calls
  * invalid ARN fails without mutating state

crdBackend (TestReconcileMappingsCRD + two focused tests):
  * deletes stale CAPA-managed CR when user removed
  * preserves out-of-band CRs (no capa-iamauth- GenerateName prefix)
  * node role preserved when included in desired set
  * idempotency (no duplicate CRs on repeated calls)
  * invalid ARN fails without mutating pre-existing CRs

Adds cmWith / readAuthCM helpers to configmap_test.go for constructing
and reading the aws-auth ConfigMap in table-driven cases.

ChellaVigneshKP committed:
refactor(iamauth): address Copilot review on PCP-2390

- reconcile.go: dedup desired role/user mappings by ARN and sort
  deterministically before calling ReconcileMappings. nodeRoles is a
  map (unordered iteration) and a user-configured mapping may collide
  with a discovered node-role ARN; without dedup+sort the ConfigMap
  backend churns aws-auth on every reconcile and the CRD backend
  creates duplicate IAMIdentityMapping CRs. User-configured mappings
  win on ARN collision (explicit intent overrides discovery).
- crd.go: scope the IAMIdentityMapping list in ReconcileMappings to
  kube-system. CAPA only ever creates these CRs there, so a
  cluster-wide list risks touching capa-prefixed CRs in other
  namespaces and unnecessarily inflates the list.
- reconcile.go: fix inconsistent klog.KRef on the final log line so it
  carries the same namespace as the entry-log line.

Adds unit tests for dedupAndSortRoles/Users covering the collision-
override and ARN-ascending-order contracts.

ChellaVigneshKP committed:
fix(iamauth): revert kube-system list scope on CRD backend

Reverts the InNamespace(metav1.NamespaceSystem) selector added in the
prior commit. IAMIdentityMapping is cluster-scoped in aws-iam-
authenticator ('+genclient:nonNamespaced' on the type,
'scope: Cluster' in the shipped CRD), so a real API server strips the
Namespace field on Create — a namespaced List returns zero items in a
real cluster and ReconcileMappings would create duplicate CRs on every
reconcile. Tests would have passed because the controller-runtime fake
client preserves the Namespace field regardless of scope.

Leaves an explanatory comment in place so this doesn't get re-suggested.

ChellaVigneshKP commented on pkg/cloud/services/iamauth/crd.go:
```diff
@@ -104,6 +113,125 @@ func (b *crdBackend) MapUser(mapping ekscontrolplanev1.UserMapping) error {
+
+	ctx := context.TODO()
+
+	mappingList := iamauthv1.IAMIdentityMappingList{}
+	if err := b.client.List(ctx, &mappingList); err != nil {
+		return fmt.Errorf("listing IAMIdentityMappings: %w", err)
+	}
```
Thanks for the suggestion — we looked at applying it but reverted the change (see [e3e6df5d4](https://github.com/spectrocloud-public/cluster-api-provider-aws/commit/e3e6df5d4)) because `IAMIdentityMapping` is **cluster-scoped**:

- `pkg/mapper/crd/apis/iamauthenticator/v1alpha1/types.go` @ v0.6.13 marks both the object and its list with `+genclient:nonNamespaced`.
- The shipped CRD (`deploy/iamidentitymapping.yaml` @ v0.6.13) has `scope: Cluster`.

On a real API server the `Namespace` field is stripped on Create for cluster-scoped resources, so listing with `crclient.InNamespace(metav1.NamespaceSystem)` would return zero items and `ReconcileMappings` would treat every reconcile as an empty starting state → duplicate CRs every loop. The controller-runtime fake client preserves the `Namespace` field regardless of scope, which is why our tests would have (falsely) passed under the namespaced-List variant.

The existing code sets `Namespace: metav1.NamespaceSystem` on Create, but that's decorative-only on the real API server. Left an inline comment in `crd.go` documenting this so it doesn't get re-suggested. The other two comments (dedup+sort of desired mappings, log-line namespace consistency) have been applied in [06e165ce5](https://github.com/spectrocloud-public/cluster-api-provider-aws/commit/06e165ce5).

cPu1 commented on pkg/cloud/services/iamauth/reconcile.go:
```diff
@@ -48,43 +51,47 @@ func (s *Service) ReconcileIAMAuthenticator(ctx context.Context) error {
+	// aws-auth backend, so omitting them would revoke kubelet auth on all
+	// worker nodes.
+	desiredRoles := make([]ekscontrolplanev1.RoleMapping, 0, len(nodeRoles)+len(iamCfg.RoleMappings))
 	for roleName := range nodeRoles {
 		roleARN, err := s.getARNForRole(roleName)
 		if err != nil {
-			return fmt.Errorf("failed to get ARN for role %s: %w", roleARN, err)
+			return fmt.Errorf("failed to get ARN for role %s: %w", roleName, err)
```
This is a genuine issue. Can we resolve this?

ChellaVigneshKP replied:
This was already addressed in 06e165ce5 (https://github.com/spectrocloud-public/cluster-api-provider-aws/pull/890/commits/06e165ce5). The map iteration at line 73 stays non-deterministic by design, but the desired slice is now passed through dedupAndSortRoles / dedupAndSortUsers (deduplicate by ARN, then sort by ARN) before ReconcileMappings — so the backend always sees a stable, unique set. User-configured mappings win on ARN collision. Unit tests in reconcile_test.go (TestDedupAndSortRoles / TestDedupAndSortUsers).
